### PR TITLE
build: make generated source files reproducible

### PR DESCRIPTION
**Pull request by raboof**
This updates the code generation scripts to be deterministic, which in turn leads to reproducible binaries.

Unfortunately this is incomplete: the exported function (funcs_metadata.generated.h) and ui event (ui_events_metadata.generated.h) metadata have some mpack'ed tables, which looks more tricky.

As a temporary workaround, this PR introduces a PRG_GEN_LUA cmake setting to specify a different lua for those parts of code generation, so you could plug in a modified build of luajit that preserves table order. This would be something to fix later.

This is a continuation of #20355

Co-Authored-By: dundargoc <gocdundar@gmail.com>
Co-Authored-By: Arnout Engelen <arnout@bzzt.net>

### CODE REVIEW

File: scripts/genvimvim.lua
Comment (justinmk) [inline scripts/genvimvim.lua:127]: similar to what's discussed in https://github.com/neovim/neovim/pull/20355#discussion_r980511837 , can the root cause be fixed instead of sorting at the "leaf logic" here?

Similar question for LUA_GEN_PRG

File: CMakeLists.txt
Comment (justinmk) [inline CMakeLists.txt:604]: If this unblocks you then perhaps it's ok as a temporary workaround, but please leave a note athttps://github.com/neovim/neovim/issues/20124 and the commit message, explaining the steps needed to unwind this temporary workaround. It looks like the steps are : 

1. fix the ordering of msgpack'd data in `funcs_metadata.generated.h` and `ui_events_metadata.generated.h`
2. ?

Comment (justinmk) [reply CMakeLists.txt:604]: btw, does this _actually_ unblock you (e.g. are you packaging for a distro?), or is this only theoretical currently?

File: scripts/genvimvim.lua
Comment (raboof) [reply scripts/genvimvim.lua]: > similar to what's discussed in [#20355 (comment)](https://github.com/neovim/neovim/pull/20355#discussion_r980511837) , can the root cause be fixed instead of sorting at the "leaf logic" here?

TBH I find it hard to say whether changing the structure of `auevents.nvim_specific` to make sure the output is deterministic is "fixing the root cause" or "leaking a presentation concern into the data layer" - but happy to adapt :)

> Similar question for LUA_GEN_PRG

Here I completely agree we should fix the 'root cause', which is the fact that the mpack'ed data contains string-keyed tables which are not serialized deterministically.

File: CMakeLists.txt
Comment (raboof) [reply CMakeLists.txt:604]: AFAICS the ordering of msgpack'd data is the only remaining work that requires this workaround. Improved the commit message, happy to also add it to #20124 when this part is merged.

> btw, does this _actually_ unblock you (e.g. are you packaging for a distro?), or is this only theoretical currently?

Yes, this helps neovim be reproducible for nixos (https://github.com/NixOS/nixpkgs/pull/208103). Of course we could also keep the patch locally in nixpkgs, but I prefer upstreaming as much as possible - so perhaps other distro's can also leverage it.

File: scripts/genvimvim.lua
Comment (justinmk) [reply scripts/genvimvim.lua]: > hard to say whether changing the structure of auevents.nvim_specific to make sure the output is deterministic is "fixing the root cause" or "leaking a presentation concern into the data layer" 

fair point, but if making the source structure deterministic has no downsides, better to do it there. Reduces entropy in general :)

File: CMakeLists.txt
Comment (justinmk) [reply CMakeLists.txt:604]: > Of course we could also keep the patch locally in nixpkgs, but I prefer upstreaming as much as possible

absolutely, we are always interested in avoiding unnecessary work for package maintainers.

File: scripts/genvimvim.lua
Comment (justinmk) [inline scripts/genvimvim.lua:99]: can `auevents.nvim_specific` be made deterministic at its source?

File: scripts/genvimvim.lua
Comment (justinmk) [reply scripts/genvimvim.lua]: > but if making the source structure deterministic has no downsides

Looking at the latest of this PR, I see a downside: having to structure e.g. `keysets.lua` as a list instead of a dict is awkward. So I agree we need something like a `sorted()` function that iterates dicts in order.

File: src/nvim/api/keysets.lua
Comment (bfredl) [inline src/nvim/api/keysets.lua:2]: While the change is the right one perhaps we can get rid of the artificial second layer of nesting:
```
{ 'name', {
  "somekey";
  "key";
}};
```

**Comment by raboof** [comment]
(huh, sorry about the review request noise, GitHub UI being weird. ready for another round AFAICS :) )

File: src/nvim/auevents.lua
Comment (justinmk) [inline src/nvim/auevents.lua:130]: re https://github.com/neovim/neovim/issues/19918 , `auevents.lua` may be a file that we want to ship in the runtime, so that plugins can discover Nvim event definitions. In that case, I'd like to keep the "map" structure of this file. So callers would need to use e.g. `spairs()` to get ordering guarantees here. @bfredl ok with this?

This doesn't apply to `keysets.lua`, which is an internal impl detail.

File: scripts/genvimvim.lua
Comment (justinmk) [inline scripts/genvimvim.lua:120]: technically this is no longer needed , right (or we could revert the changes to auevents.lua)?

Though `vim.spairs()` still seems like something we generally want.

File: runtime/lua/vim/shared.lua
Comment (justinmk) [inline runtime/lua/vim/shared.lua:481]: should we try to be more like ipairs/pairs here? From `:help luaref-pairs()` : 

```
pairs({t})                                                      *luaref-pairs()*
        Returns three values: the `next` |luaref-next()| function, the table
        {t}, and `nil`, so that the construction

               for k,v in pairs(t) do  body  end

        will iterate over all key-value pairs of table {t}.
```

i.e. should we return "iterator function, table {t}, and `nil`"? Then need some
test coverage for `for k,v in pairs(t) do  body  end`.

File: runtime/lua/vim/shared.lua
Comment (max397574) [inline runtime/lua/vim/shared.lua:470]: couldn't you use `vim.tbl_keys` here?

File: runtime/lua/vim/shared.lua
Comment (justinmk) [inline runtime/lua/vim/shared.lua:466]: In the future we could add a `sort` parameter to allow custom sorting.

File: runtime/lua/vim/shared.lua
Comment (justinmk) [inline runtime/lua/vim/shared.lua:465]:
```suggestion
---@return iterator iterator over sorted keys and their values
```

File: runtime/lua/vim/shared.lua
Comment (raboof) [reply runtime/lua/vim/shared.lua:481]: I guess so - for now I just kept it consistent with the implementation from premake. Maybe something to improve later?

File: scripts/genvimvim.lua
Comment (raboof) [reply scripts/genvimvim.lua:120]: Ah, yes, I meant to revert these changes in auevents.lua, good one. Thanks, updated.

File: runtime/lua/vim/shared.lua
Comment (justinmk) [reply runtime/lua/vim/shared.lua:481]: i suppose the door is open for changing this in the future, let's at least leave a TODO comment in the source

**Review by justinmk — APPROVED** [review]
lgtm after comments and:

- test coverage for spairs()
- do we need some sort of (small) test that verifies the reproducibility?

**Comment by raboof** [comment]
> test coverage for `spairs()`

Added a basic test

> do we need some sort of (small) test that verifies the reproducibility?

I agree that would be nice, but I'm not sure it's worth the additional infrastructure it'd introduce

**Comment by matu3ba** [comment]
> I agree that would be nice, but I'm not sure it's worth the additional infrastructure it'd introduce

Why is using `find` with computation of sha256sums + storing into a temporary file, then cleaning the generated source, recreate them and compare their sha256sums match not an option?

File: CMakeLists.txt
Comment (raboof) [reply CMakeLists.txt:266]: That's of course possible, but upstream deliberate introduced the non-determinism for security reasons (https://github.com/LuaJIT/LuaJIT/issues/626#issuecomment-707005671). While I think it's acceptable as a temporary workaround to use an 'arguably-insecure' lua for the code generation (we trust the code generator, after all), but we probably shouldn't encourage users to use such lua versions with 'more untrusted' data at runtime.

Comment (dundargoc) [reply CMakeLists.txt:266]: OK, that makes perfect sense. Thank you :)

**Comment by teto** [comment]
@raboof could you sum up what is needed before merging + possible follow up (sry I didn't foillow at all the development and the numerous comments make it hard to catch up) ? I would like to understand the situation with the nixpkgs patch (can we get rid of it after merging this ?)

File: CMakeLists.txt
Comment (justinmk) [reply CMakeLists.txt:266]: That's very good info to include in the commit message!

File: runtime/lua/vim/shared.lua
Comment (justinmk) [inline runtime/lua/vim/shared.lua:476]:
```suggestion
  -- Return the iterator function.
  -- TODO(justinmk): Return "iterator function, table {t}, and nil", like pairs()?
```

**Comment by raboof** [comment]
> could you sum up what is needed before merging

It's ready to be merged (thanks @justinmk!)

> possible follow up

The possible follow-up is described in the comment at https://github.com/neovim/neovim/blob/master/CMakeLists.txt#L245-L252

> I would like to understand the situation with the nixpkgs patch (can we get rid of it after merging this?)

Yes, we can drop the patch when neovim is released. Prepared and tested with https://github.com/NixOS/nixpkgs/pull/212235

**Pull request merged**
